algebird-core/src/main/scala/com/twitter/algebird/Batched.scala

```diff
@@ -0,0 +1,331 @@
+   * containing `this` and `that`.
+   */
+  def combine(that: Batched[T]): Batched[T] =
+    Batched.Items(this, that)
+
+  /**
+   * Compact this batch if it exceeds `batchSize`.
+   *
+   * Compacting a branch means summing it, and then storing the summed
+   * value in a new single-item batch.
+   */
+  def compact(batchSize: Int)(implicit s: Semigroup[T]): Batched[T] =
+    if (size < batchSize) this else Batched.Item(sum(s))
+
+  /**
+   * Add more values to a batched value.
+   *
+   * This method will grow the tree to the left.
+   */
+  def append(that: TraversableOnce[T]): Batched[T] =
+    that.foldLeft(this)((b, t) => b.combine(Batched(t)))
+
+  /**
+   * Provide an iterator over the underlying tree structure.
+   *
+   * This is the order used by `.sum`.
+   *
+   * This iterator traverses the tree from left-to-right. If the
+   * original expression was (w + x + y + z), this iterator returns w,
+   * x, y, and then z.
+   */
+  def iterator: Iterator[T] =
+    this match {
+      case Batched.Item(t) => Iterator.single(t)
+      case b => new Batched.ForwardItemsIterator(b)
+    }
+
+  /**
+   * Convert the batch to a `List[T]`.
+   */
+  def toList: List[T] =
+    reverseIterator.foldLeft(List.empty[T])((ts, t) => t :: ts)
+
+  /**
+   * Provide a reversed iterator over the underlying tree structure.
+   *
+   * This iterator traverses the tree from right-to-left. If the
+   * original expression was (w + x + y + z), this iterator returns z,
+   * y, x, and then w.
+   */
+  def reverseIterator: Iterator[T] =
+    this match {
+      case Batched.Item(t) => Iterator.single(t)
+      case b => new Batched.ReverseItemsIterator(b)
+    }
+
+  /**
+   * Report the size of the underlying tree structure.
+   *
+   * This is an O(1) operation -- each subtree knows how big it is.
+   */
+  def size: Int
+}
+
+object Batched {
+
+  /**
+   * Constructed a batch from a single value.
+   */
+  def apply[T](t: T): Batched[T] =
+    Item(t)
+
+  /**
+   * Constructed an optional batch from a collection of values.
+   *
+   * Since batches cannot be empty, this method returns `None` if `ts`
+   * is empty, and `Some(batch)` otherwise.
+   */
+  def items[T](ts: TraversableOnce[T]): Option[Batched[T]] =
+    if (ts.isEmpty) None else {
+      val it = ts.toIterator
+      val t0 = it.next
+      Some(Item(t0).append(it))
+    }
+
+  /**
+   * Equivalence for batches.
+   *
+   * Batches are equivalent if they sum to the same value. Since the
+   * free semigroup is associative, it's not correct to take tree
+   * structure into account when determining equality.
+   *
+   * One thing to note here is that two equivalent batches might
+   * produce different lists (for instance, if one of the batches has
+   * more zeros in it than another one).
+   */
+  implicit def equiv[A](implicit s: Semigroup[A]): Equiv[Batched[A]] =
+    new Equiv[Batched[A]] {
+      def equiv(x: Batched[A], y: Batched[A]): Boolean = x.sum(s) == y.sum(s)
```

Comment:
actually his needs `Equiv[A]` too. We should not be using universal equals inside here.